packages/node_modules/@webex/plugin-devicemanager/src/devicemanager.js

```diff
@@ -0,0 +1,503 @@
+          this.logger.error('DeviceManager#remove: failed to remove the device', error);
+          return Promise.reject(error);
+        });
+    }
+    this.logger.error('DeviceManager#remove: deviceId is required');
+    return Promise.reject(new Error('DeviceManager#remove: deviceId is required'));
+  },
+
+  /**
+  * Requests to display PIN on the device
+  * @param {object} device
+  * @param {object} options
+  * @param {object} options.data
+  * @returns {Promise<deviceInfo>}
+  */
+  requestPin(device, options = {}) {
+    const deviceId = device.id || device.identity && device.identity.id;
+    if (!deviceId) {
+      this.logger.error('DeviceManager#requestPin: device.id is required');
+      return Promise.reject(new Error('DeviceManager#requestPin: device.id is required'));
+    }
+    const space = {id: deviceId, url: `/spaces/${deviceId}`};
+    return this.spark.internal.lyra.space.get(space)
+      .then((dev) => { // check if the space is pinChallenge capable
+        if (dev && dev.occupants && dev.occupants.pinChallenge) {
+          this.logger.info('DeviceManager#requestPin: space is PIN challenge capable');
+          return this.spark.internal.lyra.space.join(space, {
+            passType: 'MANUAL',
+            verificationInitiation: 'PIN',
+            data: options.data
+          })
+            .then(() => {
+              this._devicePendingPinChallenge = dev;
+              // return the actual device so that it can be upserted on successful
+              // PIN exchange
+              return dev;
+            });
+        }
+        // pairs with the space if it's not PIN challenge capable
+        this.logger.info('DeviceManager#requestPin: space is not PIN challenge capable, probably already occupied, will still return device info');
+        return this.spark.internal.lyra.space.get(space)
+          .then(() => Promise.resolve(dev));
+      })
+      .catch((err) => {
+        this.logger.error('DeviceManager#requestPin: device failed PIN challenge request', err);
+        return Promise.reject(err);
+      });
+  },
+
+  /**
+  * pairs the device with the user (manual pairing), also adds it to
+  * user's recents list for subsequent fetches.
+  * similar to space.join()
+  * @param {object} options
+  * @param {number} options.pin
+  * @returns {Promise<deviceInfo>}
+  */
+  pair(options = {}) {
+    if (!options.pin) {
+      this.logger.error('DeviceManager#pair: options.pin is required');
+      return Promise.reject(new Error('DeviceManager#pair: options.pin is required'));
+    }
+    if (this._devicePendingPinChallenge) {
+      const space = {id: this._devicePendingPinChallenge.identity.id, url: `/spaces/${this._devicePendingPinChallenge.identity.id}`};
+
+      return this.spark.internal.lyra.space.join(space, {
+        passType: 'PIN_ANSWER',
+        data: options.pin
+      })
+        .then(() => this.upsert(this._devicePendingPinChallenge))
+        .catch((err) => {
```

Comment:
You can move this `catch` above the `then` to not catch failures that could be thrown from the upsert or update the error comment to account for a failure in either

Reply:
+1